### ISSUE TITLE
scale calculation bug (orig: "3D point cloud registration")

### ISSUE THREAD

**Issue by gamecoolha** [jewettaij/superpose3d#3]
Can you input two sets of 3D point clouds for point cloud registration?

**Comment by jewettaij** [comment]
Yes.  But the number of points in each cloud must be the same.
I think it's easier to answer this question with equations.  But I will try to use words instead.

The position of the first cloud will not change.  The Superpose3D() function will find the new position and orientation of the points in the 2nd cloud in order to minimize the sum of the square of these distances:

(distance between point 1 from cloud 1 and point 1 from cloud 2)^2 +

(distance between point 2 from cloud 1 and point 2 from cloud 2)^2 +

(distance between point 3 from cloud 1 and point 3 from cloud 2)^2 +

:

(distance between point N from cloud 1 and point N from cloud 2)^2

If the order of the points in the 2nd could does not match the order of points in the first cloud, the Superpose3D() function will not work.

I have the feeling that my answer did not address your question.  If so, feel free to clarify your question.

**Comment by gamecoolha** [comment]
Thank you for your answer,  I already know how to solve the rotation matrix and translation matrix, but I still have a question，how to solve the scaling factor?

**Comment by jewettaij** [comment]
Please forgive the slow reply.
If you set the "allow_rescale" argument to *True*, then *Superposee3D()* will attempt to find the optimal scaling factor (*c*).  You don't have to solve it yourself.

Are you asking this question because you have found a bug in the code that calculates *c*?  Is *c* not being computed correctly?

Alternatively, if you don't want to invoke *Superpose3D()* for some reason, that's okay.  If you have a different way to find the optimal rotation and position of the point clouds (without rescaling them), then use that method instead.  After doing that, calculate the radius of gyration of the two point clouds, *X*, and *x*.  You could estimate the *c* scaling parameter from the ratio of the radius of gyration for the two point clouds (assuming one is not the mirror image of the other).

**Comment by jewettaij** [comment]
> Are you asking this question because you have found a bug in the code that calculates _c_? Is _c_ not being computed correctly?

I just tried applying the code to some examples, and it appears like the answer is yes, there is a bug.  Thank you for letting me know!  My *hopeful guess* is that there is a typo in the formula used in [line 193 of the code](https://github.com/jewettaij/superpose3d/blob/a53ba6453a1b803ce837da0531eff69d83309d27/superpose3d/__init__.py#L193).
The scaling factors are off by a square-root.  That's embarrassing!  I need to go back and re-derive this formula from scratch, since it does not appear in any paper.  (I threw my notes out a long time ago.  I haven't thought about this in a while.  It looks like my automated tests are not detecting the problem either.)  I don't have time to fix this problem today.  But I'll work on it this weekend.   My guess is that I should have used this formula instead:
```
c = sqrt((WaxaiXai + pPp) / Waxaixai)
```
Meanwhile, try replacing line 193 of superpose3d/__init__.py with the line above and see if it works.  (It definitely might not.  I'm not certain if WaxaiXai + pPp>=0.)  If this approach fails, try the method I described in my previous message in this thread.

Eventually, I will fix this, either by fixing the bug, or removing the capability to calculate *c*.
Thank you very much for letting me know about this embarrassing problem!

Andrew
